### ISSUE TITLE
maps: undocumented and inconsistent behavior in the Copy function

### ISSUE THREAD

**Issue by gazerro** [golang/go#64390]
The `maps.Copy` function panics if the destination map is `nil` and the source map is not empty (`dst == nil && len(src) > 0`)

```go
var dst map[string]int
maps.Copy(dst, map[string]int(nil))    // no panic
maps.Copy(dst, map[string]int{})       // no panic
maps.Copy(dst, map[string]int{"a": 1}) // panic
```

This behavior is not currently documented.

I propose to panic if `dst` is `nil`, independently from `src`, and document it. The updated behavior would be

```go
var dst map[string]int
maps.Copy(dst, map[string]int(nil))    // panic
maps.Copy(dst, map[string]int{})       // panic
maps.Copy(dst, map[string]int{"a": 1}) // panic
```

**Comment by randall77** [comment]
It's not clear to me why the proposed behavior is any better than the current behavior. The old behavior is consistent with the built-in maps rule that nil maps panic only when written to.

I don't think we can change the behavior now. This API has already been released and people may be depending on the old behavior, even if unspecified in the docs.

**Comment by gazerro** [comment]
> The old behavior is consistent with the built-in maps rule that nil maps panic only when written to.

I would have said the opposite; I find the proposed behavior more consistent with the built-in maps rule, but now I also understand your perspective.

I'd suggest documenting the current behavior at this point; it's not quite evident without delving into the code.

**Comment by Deleplace** [comment]
I agree with most of the arguments, though IMO any usage of `Copy` with `dst==nil` is a bad idea (not a legit use case, and the developer shouldn't rely on Copy always panicking or never panicking in this case).

Keeping and documenting the existing behaviour should be fine:
```
Copy panics if dst is nil and src contains at least one element to be copied.
```

**Comment by Deleplace** [comment]
I don't think `maps.Copy` has any way to allocate memory (a new map) and provide it to the caller who passed `nil`, because `maps.Copy` doesn't return a collection like `append` or `maps.Clone`.
